language/vm/src/abi.rs

```diff
@@ -0,0 +1,399 @@
+                Vector(t) => TypeTag::Vector(Box::new(
+                    t.into_type_tag()
+                        .expect("Invariant violation: vector type argument contains reference"),
+                )),
+                Struct {
+                    address,
+                    module,
+                    name,
+                    type_arguments,
+                } => TypeTag::Struct(StructTag {
+                    address,
+                    module,
+                    name,
+                    type_params: type_arguments
+                        .into_iter()
+                        .map(|t| {
+                            t.into_type_tag().expect(
+                                "Invariant violation: struct type argument contains reference",
+                            )
+                        })
+                        .collect(),
+                }),
+                TypeParameter(_) => unreachable!(),
+            }
+        } else {
+            return None;
+        })
+    }
+}
+
+impl FieldABI {
+    /// Create a `FieldABI` for `FieldDefinition` `f` in module `m`.
+    pub fn new(m: &CompiledModule, f: &FieldDefinition) -> Self {
+        FieldABI {
+            name: m.identifier_at(f.name).to_owned(),
+            type_: TypeABI::new(m, &f.signature.0),
+        }
+    }
+}
+
+impl StructABI {
+    /// Create a `StructABI` for `StructDefinition` `def` in module `m`. Panics if `def` is a
+    /// a native struct definition.
+    pub fn new(m: &CompiledModule, def: &StructDefinition) -> Self {
+        let handle = m.struct_handle_at(def.struct_handle);
+        let fields = match &def.field_information {
+            StructFieldInformation::Native => panic!("Can't extract ABI for native struct"),
+            StructFieldInformation::Declared(fields) => {
+                fields.iter().map(|f| FieldABI::new(m, f)).collect()
+            }
+        };
+        StructABI {
+            name: m.identifier_at(handle.name).to_owned(),
+            kind: KindABI::from_bool(handle.is_nominal_resource),
+            type_parameters: handle
+                .type_parameters
+                .iter()
+                .map(|k| KindABI::new(k))
+                .collect(),
+            fields,
+        }
+    }
+}
+
+impl FunctionABI {
+    /// Create a `FunctionABI` for `FunctionHandle` `f` in module `m`.
+    pub fn new(m: &CompiledModule, f: &FunctionHandle) -> Self {
+        FunctionABI {
+            name: m.identifier_at(f.name).to_owned(),
+            type_parameters: f.type_parameters.iter().map(|k| KindABI::new(k)).collect(),
+            formals: m
+                .signature_at(f.parameters)
+                .0
+                .iter()
+                .map(|s| TypeABI::new(m, s))
+                .collect(),
+            ret: m
+                .signature_at(f.return_)
+                .0
+                .iter()
+                .map(|s| TypeABI::new(m, s))
+                .collect(),
+        }
+    }
+}
```

Comment:
The way the constructors are exposed looks very similar to the old `views` we had and that resulted in a large number of clients using the file format and the views at the same time, which was fragile and took long to clean up. With that being said this could be fine... I'm not entirely sure if the two cases are comparable, just something for us to think about.